File: Lecture25/schroedinger.py
import math
import cpt
import matplotlib.pyplot as plt

class Schroedinger : 

    def __init__(self) :
	self.hbar = 1.0                  # Planck's constant / 2pi
	self.m = 1.0                     # particle mass
	self.omega = 1.0                 # oscillator frequency
	self.E = 0.0                     # current energy in search
	self.N = 500                     # number of lattice points = N+1
	self.x_left = -5.0               # left boundary
	self.x_right = 5.0               # right boundary
	self.h = (self.x_right - self.x_left) / self.N  # grid spacing

	self.phi_left = [0.0]*(self.N+1)      # wave function integrating from left
	self.phi_right = [0.0]*(self.N+1)     # wave function integrating from right
	self.phi = [0.0]*(self.N+1)           # whole wave function

	self.sign = 1                    # current sign used to make F(E) continuous
	self.nodes = 0                   # current number of nodes in wavefunction


    def V(self, x):                   # harmonic oscillator potential
        return 0.5 * self.m * self.omega**2 * x**2


    def q(self, x):                   # Sturm-Liouville q function
        return 2 * self.m / self.hbar**2 * (self.E - self.V(x))


    def F(self, energy):              # eigenvalue at F(E) = 0

        # set energy needed by the q(x) function
        self.E = energy

        # find the right turning point
        i_match = self.N
        x = self.x_right             # start at right boundary
        while self.V(x) > self.E:         # in forbidden region
            i_match -= 1
            x -= self.h
            if i_match < 0:
                raise Exception("can't find right turning point")

        # integrate self.phi_left using Numerov algorithm
        self.phi_left[0] = 0.0
        self.phi_left[1] = 1.0e-10
        c = self.h**2 / 12.0         # constant in Numerov formula
        for i in range(1, i_match+1):
            x = self.x_left + i * self.h
            self.phi_left[i+1]  = 2 * (1 - 5 * c * self.q(x)) * self.phi_left[i]
            self.phi_left[i+1] -= (1 + c * self.q(x - self.h)) * self.phi_left[i-1]
            self.phi_left[i+1] /= 1 + c * self.q(x + self.h)

        # integrate self.phi_right
        self.phi[self.N] = self.phi_right[self.N] = 0.0
        self.phi[self.N-1] = self.phi_right[self.N-1] = 1.0e-10
        for i in range(self.N - 1, i_match - 1, -1):
            x = self.x_right - i * self.h
            self.phi_right[i-1]  = 2 * (1 - 5 * c * self.q(x)) * self.phi_right[i]
            self.phi_right[i-1] -= (1 + c * self.q(x + self.h)) * self.phi_right[i+1]
            self.phi_right[i-1] /= 1 + c * self.q(x - self.h)
            self.phi[i-1] = self.phi_right[i-1]

        # rescale self.phi_left
        scale = self.phi_right[i_match] / self.phi_left[i_match]
        for i in range(i_match + 2):
            self.phi_left[i] *= scale
            self.phi[i] = self.phi_left[i]

        # make F(E) continuous
        # count number of nodes in self.phi_left
        n = 0
        for i in range(1, i_match+1):
            if self.phi_left[i-1] * self.phi_left[i] < 0.0:
                n += 1

        # flip its sign when a new node develops

        if n != self.nodes:
            self.nodes = n
            self.sign = -self.sign

        return ( self.sign *
		 ( self.phi_right[i_match-1] - self.phi_right[i_match+1] - 
		   self.phi_left [i_match-1] + self.phi_left[i_match+1] ) /
		(2 * self.h * self.phi_right[i_match]) )

    def normalize(self):
        norm = 0.0
        for i in range(self.N):
            norm += self.phi[i]**2
        norm /= self.N
        norm = math.sqrt(norm)
        for i in range(self.N):
            self.phi[i] /= norm

print " Eigenvalues of the Schroedinger equation"
print " for the harmonic oscillator V(x) = 0.5 x^2"
print " ------------------------------------------"
E_max = input(" Enter maximum energy E: ")

phi_file = open("phi.data", "w")


level = 0                   # level number
E_old = 0.0                 # previous energy eigenvalue

schroedinger = Schroedinger()
schroedinger.E = 0.1 # guess and E below the ground state

# draw the potential
for i in range(schroedinger.N+1):
	x = schroedinger.x_left + i * schroedinger.h
swrite = '{0:12.6f} {1:12.6f}'.format( x, schroedinger.V(x) )
print swrite
print ''

# find the energy levels
print ""
print " Level       Energy           Simple Steps   Secant Steps"
print " -----   ------------------   ------------   ------------"

x_data = []
phi_data = []

while True:                 # loop over levels

	# estimate next E and dE
	dE = 0.5 * (schroedinger.E - E_old)
	E_old = schroedinger.E
	schroedinger.E += dE

	# use simple search to locate root with relatively low accuracy
	accuracy = 0.01
	schroedinger.E = cpt.root_simple(schroedinger.F, schroedinger.E, dE, accuracy)
	#simple_steps = cpt.root_steps()

	# use secant search with relatively high accuracy
	accuracy = 1.0e-6
	E1 = schroedinger.E + 100 * accuracy # guess second point required
	schroedinger.E = cpt.root_secant(schroedinger.F, schroedinger.E, E1, accuracy)
	#secant_steps = cpt.root_steps()

	#ans = " " + repr(level).rjust(3) + " "*5 + repr(E).ljust(20) + " "*6
	#ans += repr(simple_steps).rjust(3) + " "*11 + repr(secant_steps).rjust(3)
	#print ans
	level += 1

	accuracy = 0.001
	x = cpt.root_simple(schroedinger.q, schroedinger.x_left, schroedinger.h, accuracy)
	swrite = '{0:12.6f} {1:12.6f}'.format( x, schroedinger.E )
	print swrite
	
	x = cpt.root_simple(schroedinger.q, schroedinger.x_right, -schroedinger.h, accuracy)
	swrite = '{0:12.6f} {1:12.6f}'.format( x, schroedinger.E )
	print swrite

	print ''
	
	schroedinger.normalize()
	x_data.append( [] )
	phi_data.append( [] )
	iphi = len(x_data) - 1
	for i in range(schroedinger.N+1):
		x = schroedinger.x_left + i * schroedinger.h
		x_data[iphi].append(x)
		phi_data[iphi].append( schroedinger.phi[i] )


	if schroedinger.E >= E_max:          # we are done
		break




# print the search function
schroedinger.E = 0.1
dE = 0.01
E_data = []
F_data = []
while schroedinger.E < E_max:
	E_data.append( schroedinger.E )
	F_data.append(  schroedinger.F(schroedinger.E) )
	schroedinger.E += dE


s1 = plt.subplot(2,1,1)
for i in range(len(x_data)) :
	plt.plot( x_data[i], phi_data[i] )

s2 = plt.subplot(2,1,2)
plt.plot( E_data, F_data )
plt.ylim( [-10, 10] )

plt.show()

File: Lecture25/schroedinger.cpp
#include "cptstd.hpp"
#include "basalg.hpp"
#include "matrix.hpp"
using namespace cpt;

class Schroedinger {

public : 

  Schroedinger(double iE, 
	       int iN=500, double ixl = -5.0, double ixr = 5.0,
	       double iomega = 1.0,
	       double ihbar = 1.0, double im = 1.0) :
    hbar(ihbar), m(im), omega(iomega), E(iE),
    N(iN), x_left(ixl), x_right(ixr), 
    h( (x_right-x_left) / N),
    phi_left(N+1),phi_right(N+1),phi(N+1)
  {

  }

  double V(double x)                  // harmonic oscillator potential
  {
    return 0.5 * m * omega * omega * x * x;
  }


  double q(double x)                  // Sturm-Liouville q function
  {
    return 2 * m / (hbar * hbar) * (E - V(x));
  }


  double F(double iE)                  // eigenvalues at F(E) = 0
  {

    // set global E value needed by the q(x) function
    E = iE;

    // find right turning point
    int i_match = N;
    double x = x_right;             // start at right boundary
    while (V(x) > E)  {             // in forbidden region
      --i_match;
      x -= h;
      if (i_match < 0) {
	cerr << "can't find right turning point" << endl;
	exit(EXIT_FAILURE);
      }
    }

    // integrate phi_left using Numerov algorithm
    phi_left[0] = 0;
    phi_left[1] = 1e-10;
    double c = h * h / 12;          // constant in Numerov formula
    for (int i = 1; i <= i_match; i++) {
      x = x_left + i * h;
      phi_left[i+1]  = 2 * (1 - 5 * c * q(x)) * phi_left[i];
      phi_left[i+1] -= (1 + c * q(x - h)) * phi_left[i-1];
      phi_left[i+1] /= 1 + c * q(x + h);
    }

    // integrate phi_right
    phi[N]   = phi_right[N]   = 0;
    phi[N-1] = phi_right[N-1] = 1e-10;
    for (int i = N - 1; i >= i_match; i--) {
      x = x_right - i * h;
      phi_right[i-1]  = 2 * (1 - 5 * c * q(x)) * phi_right[i];
      phi_right[i-1] -= (1 + c * q(x + h)) * phi_right[i+1];
      phi[i-1] = phi_right[i-1] /= 1 + c * q(x - h);
    }

    // rescale phi_left
    double scale = phi_right[i_match] / phi_left[i_match];
    for (int i = 0; i <= i_match + 1; i++)
      phi[i] = phi_left[i] *= scale;

    // make F(E) continuous
    static int sign = 1;            // current sign used
    static int nodes = 0;           // current number of nodes

    // count number of nodes in phi_left
    int n = 0;
    for (int i = 1; i <= i_match; i++)
      if (phi_left[i-1] * phi_left[i] < 0)
	++n;

    // flip its sign when a new node develops
    if (n != nodes) {
      nodes = n;
      sign = -sign;
    }

    return sign * ( phi_right[i_match-1] - phi_right[i_match+1]
		    - phi_left[i_match-1] + phi_left[i_match+1] )
      / (2 * h * phi_right[i_match]);
  }

  void normalize() {
    double norm = 0;
    for (int i = 0; i < N; i++)
      norm += phi[i] * phi[i];
    norm /= N;
    norm = sqrt(norm);
    for (int i = 0; i < N; i++)
      phi[i] /= norm;
  }

  void set_E(double iE) { E = iE;}

  int get_N() const { return N;}
  double get_x_left() const { return x_left;}
  double get_x_right() const { return x_right;}
  double get_h() const { return h; }

  double get_phi(int i) const { return phi[i]; }

protected : 

  double hbar ;               // Planck's constant / 2pi
  double m ;                  // particle mass
  double omega ;              // oscillator frequency

  double E;                   // current energy in search

  int N ;                     // number of lattice points = N + 1
  double x_left;              // left boundary
  double x_right;             // right boundary
  double h ;                  // grid spacing

  Matrix<double,1> phi_left;  // wave function integrating from left
  Matrix<double,1> phi_right; // wave function integrating from right
  Matrix<double,1> phi;       // whole wavefunction


};

class ScroedingerWrapFE  {
public : 
  ScroedingerWrapFE( Schroedinger & inputS) : s(inputS) {}

  double operator()( double x ) {
    return s.F(x);
  }

protected : 
  Schroedinger & s;

};


class ScroedingerWrapQ  {
public : 
  ScroedingerWrapQ( Schroedinger & inputS) : s(inputS) {}

  double operator()( double x ) {
    return s.q(x);
  }

protected : 
  Schroedinger & s;

};

int main() {
    cout << " Eigenvalues of the Schroedinger equation\n"
         << " for the harmonic oscillator V(x) = 0.5 x^2\n"
         << " ------------------------------------------\n"
         << " Enter maximum energy E: ";
    double E_max;
    cin >> E_max;


    // find the energy levels
    cout << "\n Level       Energy       Simple Steps   Secant Steps"
         << "\n -----   --------------   ------------   ------------\n";
    int level = 0;                      // level number
    double E_old = 0;                   // previous energy eigenvalue
    double E = 0.1;                            // guess an E below the ground state

    Schroedinger s(E);                 // Main Schroedinger object
    // Note : these two are holding active, non-const references to s! 
    ScroedingerWrapFE s_wrapfe( s );   // Wrapper class to call Schroedinger::F(E) from operator()
    ScroedingerWrapQ s_wrapq( s );     // Wrapper class to call Schroedinger::q(E) from operator()



    ofstream levels_file("levels.data");
    ofstream phi_file("phi.data");

    // draw the potential
    for (int i = 0; i <= s.get_N(); i++) {
      double x = s.get_x_left() + i * s.get_h();
      levels_file << x << '\t' << s.V(x) << '\n';
    }
    levels_file << '\n';

    do {


        // estimate next E and dE
        double dE = 0.5 * (E - E_old);
        E_old = E;
        E += dE;

	s.set_E( E );   // Remember, this is referenced below by s_wrapfe and s_wrapq. 


        // Simple search to locate next energy level approximately
        SimpleSearchT<ScroedingerWrapFE>  simpleFE;
        double accuracy = 0.01;         // set a relatively low accuracy
        simpleFE.set_accuracy(accuracy);
        simpleFE.set_first_root_estimate(E);
        simpleFE.set_step_estimate(dE);
        E = simpleFE.find_root(s_wrapfe);

        // Secant search to locate energy level precisely
        accuracy = 0.000001;            // can use a relatively high accuracy
        double E1 = E + 100 * accuracy; // guess second point required by secant
        SecantSearchT<ScroedingerWrapFE> secant;
        secant.set_accuracy(accuracy);
        secant.set_first_root_estimate(E);
        secant.set_step_estimate(dE);
	s.set_E(E);
        E = secant.find_root(s_wrapfe);


        // Simple search to locate zeros of Schroedinger::q
        SimpleSearchT<ScroedingerWrapQ>  simpleQ;
        simpleQ.set_accuracy(accuracy);

	
        cout.precision(10);
        cout << setw(4) << level++ << "  "
             << setw(15) << E << "  " << endl;
        simpleQ.set_first_root_estimate(s.get_x_left());
        simpleQ.set_step_estimate(s.get_h());
        levels_file << simpleQ.find_root(s_wrapq) << '\t' << E << '\n';
        simpleQ.set_first_root_estimate(s.get_x_right());
        simpleQ.set_step_estimate(-s.get_h());
        levels_file << simpleQ.find_root(s_wrapq) << '\t' << E << '\n';
        levels_file << '\n';
        s.normalize();
        for (int i = 0; i <= s.get_N(); i++) {
	  double x = s.get_x_left() + i * s.get_h();
	  phi_file << x << '\t' << s.get_phi(i) << '\n';
        }
        phi_file << '\n';
    } while (E < E_max);

    levels_file.close();
    phi_file.close();

    // output the search function to a file
    ofstream search_file("F.data");
    E = 0.1;
    double dE = 0.01;
    while (E < E_max) {
        search_file << E << '\t' << s.F(E) << '\n';
        E += dE;
    }
    search_file.close();

    cout << "\n Energy levels in file levels.data"
         << "\n Eigenfunctions in file phi.data"
         << "\n Search function in file F.data" << endl;
}
